h2/src/main/org/h2/mvstore/tx/Transaction.java

```diff
@@ -664,16 +664,17 @@ private void notifyAllWaitingTransactions() {
      * @param toWaitFor transaction to wait for
      * @param mapName name of the map containing blocking entry
      * @param key of the blocking entry
+     * @param timeoutMillis timeout in milliseconds, {@code -1} for default
      * @return true if other transaction was closed and this one can proceed, false if timed out
      */
-    public boolean waitFor(Transaction toWaitFor, String mapName, Object key) {
+    public boolean waitFor(Transaction toWaitFor, String mapName, Object key, int timeoutMillis) {
         blockingTransaction = toWaitFor;
         blockingMapName = mapName;
         blockingKey = key;
         if (isDeadlocked(toWaitFor)) {
             tryThrowDeadLockException(false);
         }
-        boolean result = toWaitFor.waitForThisToEnd(timeoutMillis, this);
+        boolean result = toWaitFor.waitForThisToEnd(timeoutMillis < 0 ? this.timeoutMillis : timeoutMillis, this);
```

Comment:
rather use timeoutMillis == -1 here, since there is only one valid "not normal" value